### PR TITLE
3874 use several cells to index

### LINKED ISSUE #3874: Booster: More flexible rewrite rule indexing
The rule index implementation is currently a simple hard-wired search for a cell called `k` containing a non-empty `kseq` ). 

We can obviously do better by 
* making this cell (`k` by default) configurable for the server
* adding a way to consider more than one cell for the index. This is important for layered semantics. As an example: for the MX backend, we would need `instrs`, `k`, and  `commands`). This alternative is still an over-approximation but an improvement over picking just one of these three important cells.

Computing this automatically as suggested in #3769 would be a great addition to booster, and can happen at load time or at compile time (adding metadata to the definition). 
As a first step to address current issues, we can quickly implement the manual mode described above.

### PR DESCRIPTION
**Pull request by jberthold**
* The index on the `k` cell is generalised to an index on one or more arbitrary cells containing K sequences (default stays the same, the `k` cell only). 
* The cell(s) to index are read from a definition attribute named `indexCells`. This attribute is expected to contain an argument list of names _quoted as all symbols in kore_  (e.g., `Lbl'-LT-'k'-GT-'` for the `k` cell).
* Alternatively, the cell(s) to index can be overridden by a command line option `--index-cells CELL1,CELL2,...` to `booster-dev` and `kore-rpc-booster`.  This option takes plain names (like `k` for the `k` cell), which are quoted and prefixed before overriding the definition attribute of the loaded definition.

Fixes #3874

### CODE REVIEW

**Comment by Baltoli** [comment]
When this is merged we can look at getting support for the `indexCells` attribute added to the frontend. I suspect what we'll want to do is:
- Allow `index` as a K cell attribute
- Add a pass before generating KORE that pulls `index` off each configuration cell and composes the correctly-formatted `indexCells` attribute for the definition.

**Comment by ehildenb** [comment]
Can you post information about how much this improves the mx-semantics repository verification capabilities?

**Comment by jberthold** [comment]
| Test                                   | 3874-use-several-cells-to-index time | master-e62fba2f0 time | (3874-use-several-cells-to-index/master-e62fba2f0) time 
|----------------------------------------|--------------------------------------|-----------------------|---------------------------------------------------------
| erc20/hkg/totalSupply-spec.k           | 35.34                                | 43.83                 | 0.806297056810404                                       
| mcd/vat-addui-fail-rough-spec.k        | 56.67                                | 69.74                 | 0.8125896185833095                                      
| erc20/hkg/transfer-failure-1-spec.k    | 45.77                                | 54.29                 | 0.8430650211825383                                      
| mcd/end-cash-pass-rough-spec.k         | 186.62                               | 214.96                | 0.868161518422032                                       
| erc20/hkg/transfer-failure-2-spec.k    | 45.66                                | 52.56                 | 0.8687214611872145                                      
| erc20/hkg/transfer-success-1-spec.k    | 41.13                                | 46.83                 | 0.8782831518257528                                      
| mcd/vat-addui-pass-spec.k              | 93.52                                | 103.06                | 0.9074325635552105                                      
| mcd/end-pack-pass-rough-spec.k         | 129.96                               | 141.78                | 0.9166314007617435                                      
| erc20/hkg/balanceOf-spec.k             | 36.07                                | 39.26                 | 0.9187468160978095                                      
| mcd/pot-join-pass-rough-spec.k         | 166.61                               | 177.79                | 0.9371168232184038                                      
| erc20/hkg/transfer-success-2-spec.k    | 40.39                                | 41.94                 | 0.9630424415832142                                      
| erc20/ds/transferFrom-success-1-spec.k | 65.58                                | 63.29                 | 1.0361826512877232                                      
| mcd/vat-subui-pass-rough-spec.k        | 96.87                                | 90.61                 | 1.0690872972078138                                      
| TOTAL                                  | 1040.19                              | 1139.94    | 0.9124953944944473

**Comment by jberthold** [comment]
> Can you post information about how much this improves the mx-semantics repository verification capabilities?

I ran some tests locally using  `--index-cells instrs,commands,k`:

| Test                          | index `instrs,commands,k` | default `k` |
|-------------------------------|----------------|--------|
| test_add_liquidity-spec.json  |          13965 |  15515 |
| test_change_quorum-spec.json  |           2206 |   2764 |
| test_call_add-spec.json       |            126 |    147 |
| test_call_add_twice-spec.json |            204 |    242 |
| test_fund-spec.json           |            188 |    220 |

These runs were from an `mx-backend` branch that got merged today. 
The `add_liquidity` "default" run was from yesterday, without latest `mx-backend` tweaks.

**Comment by ehildenb** [comment]
Awesome, thanks. Have you tried different orders like `commands,instrs,k`, or `commands,k,instrs`?

File: booster/library/Booster/Definition/Attributes/Base.hs [hunk 1]
```diff
@@ -64,14 +65,15 @@ import Booster.SMT.Base (SExpr)
 import Booster.Util (Flag (..))
 import Booster.Util qualified as Util
 
-data DefinitionAttributes = DefinitionAttributes
-    {
+newtype DefinitionAttributes = DefinitionAttributes
+    { indexCells :: Maybe [ByteString] -- which cells to index for rewriting
```
Comment (geo2a) [inline booster/library/Booster/Definition/Attributes/Base.hs:69]: Perhaps worth making it a non-empty list, as we will always have a least one cell to index?

Comment (jberthold) [reply booster/library/Booster/Definition/Attributes/Base.hs:69]: True, running without any index cell is probably not a use case (though possible now).
We could model this by taking the empty list to the default `k` index.

**Comment by jberthold** [comment]
> Awesome, thanks. Have you tried different orders like `commands,instrs,k`, or `commands,k,instrs`?

I chose this order because `instrs` is the most discriminating cell (leads to the least rules with index `Anything` when used in isolation). However, the order actually does not matter for that (all cell indexes are used equally when selecting the rules for a term).

**Comment by jberthold** [comment]
These are the `kontrol` test timings, basically just noise IMHO (parallelism = 3)

| Test                                                                          | 3874-use-several-cells-to-index time | master-e62fba2f0 time | (3874-use-several-cells-to-index/master-e62fba2f0) time 
|-------------------------------------------------------------------------------|--------------------------------------|-----------------------|---------------------------------------------------------
| EmitContractTest.testExpectEmitDoNotCheckData()                               | 65.13                                | 75.15                 | 0.8666666666666666                                      
| ExpectRevertTest.test_expectRevert_true()                                     | 64.07                                | 68.81                 | 0.931114663566342                                       
| ExpectRevertTest.testFail_expectRevert_empty()                                | 28.76                                | 30.07                 | 0.9564349850349185                                      
| ExpectRevertTest.test_expectRevert_returnValue()                              | 115.5                                | 111.57                | 1.0352245227211616                                      
| MockCallTestFoundry.testMockCallEmptyAccount()                                | 59.09                                | 57.05                 | 1.0357581069237511                                      
| ExpectRevertTest.test_expectRevert_bytes4()                                   | 68.52                                | 66.1                  | 1.0366111951588504                                      
| PlainPrankTest.test_prank_zeroAddress_true()                                  | 101.17                               | 97.57                 | 1.0368965870656965                                      
| FeeTest.test_fee_setup()                                                      | 43.47                                | 41.61                 | 1.0447007930785868                                      
| ExpectRevertTest.test_expectRevert_message()                                  | 71.44                                | 68.26                 | 1.0465865807207735                                      
| StartPrankTestMsgSender.test_startprank_msgsender_setup()                     | 74.84                                | 71.41                 | 1.0480324884469963                                      
| test_foundry_prove.py::test_foundry_merge_nodes | 30.33                                | 28.93                 | 1.0483926719668164                                      
| ExpectCallTest.testExpectRegularCall()                                        | 66.94                                | 63.68                 | 1.0511934673366834                                      
| ExpectCallTest.testExpectStaticCall()                                         | 52.61                                | 49.56                 | 1.0615415657788538                                      
| test_foundry_prove.py::test_foundry_xml_report  | 17.79                                | 16.53                 | 1.0762250453720508                                      
| EmitContractTest.testExpectEmitLessTopics()                                   | 72.39                                | 65.83                 | 1.099650615221024                                       
| ExpectRevertTest.testFail_expectRevert_bytes4()                               | 72.51                                | 65.68                 | 1.1039890377588306                                      
| TOTAL                                                                         | 1004.56                              | 977.81     | 1.027357053006208

File: booster/library/Booster/Definition/Attributes/Base.hs (hunk 1 above)
Comment (jberthold) [reply booster/library/Booster/Definition/Attributes/Base.hs]: I implemented the interpretation of `[]` as "use the default" now.

**Review by geo2a — APPROVED** [review]
Great work, I'm happy to merge!

**Pull request merged**